File: BOJ_prob/Greedy/BOJ_2873.py
#*
import sys
from collections import deque
def findMin():
    minvalidx = (0, 0)
    minval = 1000
    global r, c
    for i in range(r):
        for j in range(c):
            if not (i == 0 and j == 0) and not (i == r -1 and j == c -1):
                if (i % 2 == 0 and j % 2 != 0) or (i % 2 != 0 and j % 2 == 0):
                    if groundinfo[i][j] < minval:
                        minval = groundinfo[i][j]
                        minvalidx = (i, j)
    return minvalidx

def inRange(cur_r, cur_c):
    if 0 <= cur_r < r and 0 <= cur_c < c:
        return True
    return False


def getPath0(minr, minc):
    global r, c
    path.append(('D' * (r -1) + 'R' + 'U' * (r - 1) + 'R') * (minc//2))
    cur_r, cur_c = 0, 2 * (minc//2)
    cbound = cur_c + 1
    while not (cur_c == cbound and cur_r == r -1):
        if cur_c < cbound and [cur_r, cur_c + 1] != [minr, minc]:
            path.append('R')
            cur_c += 1
        elif cur_c == cbound  and [cur_r, cur_c - 1] != [minr, minc]:
            path.append('L')
            cur_c -= 1
        if cur_r + 1 <= r - 1:
            path.append('D')
            cur_r += 1
    path.append(('R' + 'U' * (r - 1) + 'R' + 'D' * (r - 1)) * ((c - minc - 1) // 2))
def getPath1():
    global r, c
    for i in range(c):
        if i % 2 == 0:
            path.append('D' * (r - 1))
            if i != c - 1:
                path.append('R')
        else:
            path.append('U' * (r - 1))
            path.append('R')

def getPath2():
    global r, c
    for i in range(r):
        if i % 2 == 0:
            path.append('R' * (c - 1))
            if i != r - 1:
                path.append('D')
        else:
            path.append('L' * (c - 1))
            path.append('D')

r, c = map(int, sys.stdin.readline().split())
groundinfo = []
path = []
for i in range(r):
    groundinfo.append(list(map(int, sys.stdin.readline().split())))
if r % 2 == 0 and c % 2 == 0:
    minr, minc = findMin()
    groundinfo[minr][minc] = 0
    getPath0(minr, minc)
elif c % 2 != 0 and r % 2 == 0:
    getPath1()
else:
    getPath2()
for i in path:
    print(i, end='')

File: BOJ_prob/Greedy/C++/BOJ_2873.cpp
//*
#include <iostream>
#include <vector>
#include <string>
using namespace std;

int r, c, minr, minc;
int map[1000][1000];
string result;

void traverse1();
void traverse2();
void removeMin();
void traverse3();

int main()
{
    cin >> r >> c;
    for(int i = 0; i < r; i++){
        for(int j = 0; j < c; j++){
            cin >> map[i][j];
        }
    }
    if((r % 2 != 0 && c % 2 != 0) || (r % 2 != 0 && c % 2 == 0)){
        traverse1();
    }
        
    else if(r % 2 == 0 && c % 2 != 0){
        traverse2();
    }
        
    else{
        removeMin();
        traverse3();
    }
    cout << result << endl;
}

void traverse1()
{
    for(int i = 0; i < r; i++){
        if(i % 2 == 0){
            for(int j = 0; j < c - 1; j++){
                result+=("R");
            }
        }
        else{
            for(int j = 0; j < c - 1; j++){
                result+=("L");
            }
        }
        if(i != r - 1)
            result+=("D");
    }
}

void traverse2()
{
    for(int i = 0; i < c; i++){
        if(i % 2 == 0){
            for(int j = 0; j < r - 1; j++){
                result+=("D");
            }
        }
        else{
            for(int j = 0; j < r - 1; j++){
                result+=("U");
            }
        }
        if(i != c - 1)
            result+=("R");
    }
}

void removeMin()
{
    
    int minval = 987654321;
    for(int i = 0; i < r; i++){
        for(int j = 0; j < c; j++){
                if((i % 2 == 0 && j % 2 != 0) || (i % 2 != 0 && j % 2 == 0)){
                if(map[i][j] < minval){
                    minr = i;
                    minc = j;
                    minval = map[i][j];
                }
            }  
        }
    }
}

void traverse3()
{
    for(int i = 0; i < minc /2 ; i++){
        for(int j = 0; j < r - 1; j++){
            result+=("D");
        }
        result+=("R");
        for(int j = 0; j < r - 1; j++){
            result+=("U");
        }
        result+="R";
    }
    int cur_r = 0, cur_c = minc / 2 * 2;
    int max_c = cur_c + 1;
    while(!(cur_r == r - 1 && cur_c == max_c)){
        if(!(cur_r == minr && cur_c + 1 == minc) && cur_c < max_c){
            result+=("R");
            cur_c++;
        }
        else if(cur_c == max_c && !(cur_r == minr && cur_c - 1 == minc)){
            result+=("L");
            cur_c--;
        }
        if(cur_r < r - 1){
            result+=("D");
            cur_r++;
        }
    }

    for(int i = 0; i < (c - minc - 1) / 2; i++){
        result+=("R");
        for(int j = 0; j < r - 1; j++){
            result+=("U");
        }
        result+=("R");
        for(int j = 0; j < r - 1; j++){
            result+=("D");
        }
    }
    
}
